### DISCUSSION TITLE
Connecting Panda arm in gazebo to another object/robot

### DISCUSSION THREAD

**Discussion by BenOSaunders** [justagist/panda_simulator#44] [Q&A]
We are using this to make a simulation for some research involving a Clearpath Husky and a Panda arm and am running into issues. We can't figure out how to connect/mount the Panda arm to another object. We made a launch file to launch both the Husky and the Panda into the same gazebo world. We also modified the franka_panda_description URDF to use the connected_to parameter already in it to connect the Panda to a Husky link (in the Husky URDF) but it crashes due to this. Any ideas or experience getting this Panda simulation mounted onto another object in gazebo? Or examples on how to utilize the connected_to parameter in the franka_panda_description urdf?

**Comment by justagist** [comment]
Hi,
Since you only say that it crashed and don't specify any log output or error messages, I can't say for sure what you have tried so far, but I'll try to give you some tips based on the (very little) experience I have with working on a non-'world' base frame for the robot in Gazebo.

First of all, you need to make sure that the TF tree is complete (i.e. the root node is `world` frame), when you define a custom frame to be used as the robot's base. It may be possible to avoid doing this for mobile robots but I'm not very sure how. But more importantly, each link should be explicitly defined in the urdf for gazebo.

As a simple example on how to use the 'connected_to' parameter, you can look at the attached zip file (
[panda_arm_xacro_mods.zip](https://github.com/justagist/panda_simulator/files/6478195/panda_arm_xacro_mods.zip)). This is just a modified version of the panda_arm_hand.urdf.xacro file from the 'franka_panda_description' package to show a **very simplified** additional element in gazebo's tf structure. You can try replacing these file in the `franka_panda_description/robots/` directory and launch the simulator to see that the robot now sits on top of a block (see image below). The only change required for adding a base is in line 7 of panda_arm_hand.urdf.xacro file (to specify the frame for the base to connect to), and in line 3 of panda_arm.xacro file (to specify the origin of the joint to fix the robot; note that this can also be directly set in panda_arm_hand.urdf.xacro file by passing it as another parameter, which means you only have to change values in one file). *In the attached panda_arm.xacro file, I have also added lines 10 to 39, but this is simply defining the block and the joint connecting it to the world; this should actually be done in your other urdf file where the base is defined.*

![pgz_base_demo](https://user-images.githubusercontent.com/22923089/118259943-aa072600-b4a9-11eb-8148-01788ebf929b.png)

I realise your case may be more complicated because of the more complex structure and the mobile nature of the base; I have not tested it with a moving base, and you may need to do more than just what I have explained above. 

Also, you may find it useful to go through the [sawyer_simulator](https://github.com/RethinkRobotics/sawyer_simulator) and [sawyer_description](https://github.com/RethinkRobotics/sawyer_robot/tree/master/sawyer_description) packages; they have the functionality to mount the robot on a more complex structure (pedestal) and it looks like they have a way of making the base non-static (I have not tried this feature out myself; I have only gone through their description package quickly).

I hope this helps you in some way. If you find a better way to incorporate adding a different base for the robot, please let me know and/or feel free to raise a PR!

**Comment by justagist** [comment]
I just saw that you had posted [this question in the Franka Community forum](https://www.franka-community.de/t/panda-gazebo-mount-on-external-object/2972) (I am assuming this is you because the question is similar). Please note that this repository is not an official package by Franka Emika, so they will not be able to provide you with any support for this package or any of my other repositories.
These repositories are purely personal projects of mine that I developed (with a lot of help from many awesome people from the open-source community), mainly as part of my research, and also for learning to develop reusable software and interfaces.
